Tests/Functional/Drivers/Queue/PhpAmqpLibDriverTest.php

```diff
@@ -0,0 +1,244 @@
+        $this->assertNull($return);
+    }
+
+    /**
+     * Tests the consume process with a callback in return that receives the messages and ack this after
+     *
+     * @dataProvider getMessages
+     * @group amqp-consume-callback
+     */
+    public function testConsumeWithCallbackAckingMessage(MessageInterface $msg)
+    {
+        $this->prepareToConsume($msg);
+
+        $amqpMessage = null;
+        $callback = function(AMQPMessage $message) use (&$amqpMessage){
+            $amqpMessage = $message;
+            $queueMessage = new QueueMessage();
+            $queueMessage->setMessageId($amqpMessage->getDeliveryTag());
+            $this->driver->ack($queueMessage);
+        };
+
+        $this->driver->consume($this->consumer->getName(), $this->queueName, $callback);
+        $this->driver->isConsuming();
+        $this->driver->wait();
+
+        $this->assertInstanceOf(AMQPMessage::class, $amqpMessage);
+        $this->assertEquals($amqpMessage->delivery_info['routing_key'], $this->queueName);
+        $this->assertInstanceOf(AsyncQueueConsumer::class, $this->consumer);
+        $this->assertEquals($amqpMessage->delivery_info['consumer_tag'], $this->consumer->getName());
+        $this->assertContains('QueueMessage', $amqpMessage->getBody());
+    }
+
+    /**
+     * Tests the consume process with a callback in return that receives the messages and nack this after
+     *
+     * @dataProvider getMessages
+     * @group amqp-consume-callback-nack
+     */
+    public function testConsumeNackingMessage(MessageInterface $msg)
+    {
+        $this->prepareToConsume($msg);
+        $amqpMessage = null;
+
+        $callback = function($message) use (&$amqpMessage) {
+            $amqpMessage =  $message;
+            $queueMessage = new QueueMessage();
+            $queueMessage->setMessageId($amqpMessage->getDeliveryTag());
+            $this->driver->nack($queueMessage);
+        };
+
+        $this->driver->consume($this->consumer->getName(), $this->queueName, $callback);
+        $this->driver->isConsuming();
+        $this->driver->wait();
+
+        $this->assertInstanceOf(AMQPMessage::class, $amqpMessage);
+        $this->assertEquals($amqpMessage->delivery_info['routing_key'], $this->queueName);
+        $this->assertInstanceOf(AsyncQueueConsumer::class, $this->consumer);
+        $this->assertEquals($amqpMessage->delivery_info['consumer_tag'], $this->consumer->getName());
+        $this->assertContains('QueueMessage', $amqpMessage->getBody());
+    }
+
+    /**
+     * Tests the function to destroy the connection and clean the variables related
+     *
+     * @group destroy
+     */
+    public function testDestroy()
+    {
+        $this->driver->connect();
+        $this->driver->declareChannel();
+        $this->assertTrue($this->driver->isConnected());
+        $this->driver->destroy($this->createConsumer()->getName());
+        $this->assertFalse($this->driver->isConnected());
```

Comment:
To test the destroy correctly shouldn't we have the driver consuming? With what it is now implemented now in the test there is a channel without consume

Reply:
I didn't got you. The channel does not have a consumer tag once it belongs to the consumer and the at this point the consumer has  a name (tag).

Reply:
Let me try to explain better. The method destroy requires the consumer tag as parameter. Internally what does is 
```
$this->channel->basic_cancel($consumerTag);
```
My question is, doesn't need to happen
```
$this->channel->basic_consume($queueName, $consumerTag, false, false, false, false, $callback);
```
before than executing the destroy?

Reply:
Ok, I got it. But, I don't see any impact of applying this or not once the result will be the same. In any case, I'll apply the consume before the destroy.

Reply:
If we want to make sure the destroy method works, we need to check that the queue consumer ends, and for that we need to start consuming before

Reply:
Alright. I applied the consume before the destroy call.